cpp/test/cluster/kmeans_balanced.cu

```diff
@@ -0,0 +1,235 @@
+                                          p.n_rows,
+                                          p.n_cols,
+                                          p.n_clusters,
+                                          stream,
+                                          true,
+                                          nullptr,
+                                          nullptr,
+                                          MathT{0.1},
+                                          true,
+                                          MathT{-1},
+                                          MathT{1},
+                                          (uint64_t)1234);
+
+    // Convert blobs dataset to DataT if necessary
+    if constexpr (!std::is_same_v<DataT, MathT>) {
+      raft::linalg::unaryOp(
+        X.data_handle(), blobs.data(), p.n_rows * p.n_cols, op.reverse_op, stream);
+    }
+
+    d_labels.resize(p.n_rows, stream);
+    d_labels_ref.resize(p.n_rows, stream);
+    d_centroids.resize(p.n_clusters * p.n_cols, stream);
+
+    raft::linalg::unaryOp(
+      d_labels_ref.data(), blob_labels.data_handle(), p.n_rows, raft::cast_op<LabelT>(), stream);
+
+    auto X_view =
+      raft::make_device_matrix_view<const DataT, IdxT>(X.data_handle(), X.extent(0), X.extent(1));
+    auto d_centroids_view =
+      raft::make_device_matrix_view<MathT, IdxT>(d_centroids.data(), p.n_clusters, p.n_cols);
+    auto d_labels_view = raft::make_device_vector_view<LabelT, IdxT>(d_labels.data(), p.n_rows);
+
+    raft::cluster::kmeans_balanced::fit_predict(
+      handle, p.kb_params, X_view, d_centroids_view, d_labels_view, op);
+
+    handle.sync_stream(stream);
+
+    score = raft::stats::adjusted_rand_index(
+      d_labels_ref.data(), d_labels.data(), p.n_rows, handle.get_stream());
+
+    if (score < 1.0) {
+      std::stringstream ss;
+      ss << "Expected: " << raft::arr2Str(d_labels_ref.data(), 25, "d_labels_ref", stream);
+      std::cout << (ss.str().c_str()) << '\n';
+      ss.str(std::string());
+      ss << "Actual: " << raft::arr2Str(d_labels.data(), 25, "d_labels", stream);
+      std::cout << (ss.str().c_str()) << '\n';
+      std::cout << "Score = " << score << '\n';
+    }
+  }
+
+  void SetUp() override { basicTest(); }
+
+ protected:
+  raft::handle_t handle;
+  cudaStream_t stream;
+  rmm::device_uvector<LabelT> d_labels;
+  rmm::device_uvector<LabelT> d_labels_ref;
+  rmm::device_uvector<MathT> d_centroids;
+  double score;
+};
+
+template <typename MathT, typename IdxT>
+std::vector<KmeansBalancedInputs<MathT, IdxT>> get_kmeans_balanced_inputs()
+{
+  std::vector<KmeansBalancedInputs<MathT, IdxT>> out;
+  KmeansBalancedInputs<MathT, IdxT> p;
+  p.kb_params.n_iters = 20;
+  p.kb_params.metric  = raft::distance::DistanceType::L2Expanded;
+  p.tol               = MathT{0.0001};
+  std::vector<std::tuple<size_t, size_t, size_t>> row_cols_k = {
+    {1000, 32, 5}, {1000, 100, 20}, {10000, 32, 10}, {10000, 100, 50}, {10000, 500, 100}};
```

Comment:
Would you consider adding a single or couple tests for the memory footprint? e.g. set a small number of clusters and training iterations, but the size in the orders of magnitude of the GPU memory. To have at some understanding of the out-of-memory error limits.

Reply:
I've added a test with more rows but I can't push it close to the memory limits, as the tests are supposed to run on smaller GPUs as well. We can always edit these parameters manually when testing for this locally.